### PR TITLE
a bit of type checking

### PR DESCRIPTION
**Pull request by Tcharl**
Move yo-rc definition to a typescript object.

Further PR will try to narrow the amount of fields filled at each step of generator phases and move the types to their respective generators. Not perfect yet and waiting for advices

contribute to https://github.com/jhipster/generator-jhipster/issues/26114

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mshima** [comment]
I don't think we should make `jhipsterConfig` typed.
- blueprints will need to convert to any before using.

**Comment by Tcharl** [comment]
This is why I kept the `& Record<string, any>` at the end of the `AbstractJSONGeneratorJhipsterContent`.
I'm convinced that knowing all the available and filled options at every step of every generator will improve readability, velocity, stability and keep (once everything done) cli and jdl options aligned.

Please change your mind: it's definitely hard to introduce (or fix) the current codebase for not near full time contributors, and setting type checking would definitely help in that regard (at least).

**Comment by mshima** [comment]
`jhipsterConfig` and `jhipsterConfigWithDefaults` are never manually used anymore.

Command has evolved as somewhat source of truth.
https://github.com/mshima/generator-jhipster/blob/f0a8e84e2fbd24ab2f7e9cbe2bcac581004aa671/generators/git/command.ts#L37-L39

A storage scoped config:
- allows cli option/argument with choices
- allows prompts
- allows post processing 
- is loaded from cli/prompt to the jhipsterConfig
- is loaded from jhipsterConfigWithDefaults to the application object.

Probably will allow jdl definitions in the future.

**Comment by Tcharl** [comment]
Thanks for the tips, seems to be a very nice approach

Still, `jhipsterConfig*` are used in the codebase (via the 'loaded from the cli to jhipster config') so from a codebase point of view using types should help the contributors to understand features more easily. I plan to use the same typing for `application` and to merge the `jdl/JSONGeneratorJhipsterXContent` with their `<generator>/types.d` content in the next commits: i.e. removing `jdl/JSONGeneratorJhipsterAuthenticationContent` in favor of `base-application/types.d/AuthenticationType` to avoid code dupe and enforce consistency.

Just doing it step by step

**Comment by Tcharl** [comment]
Is it correct? simplifies a bit the default value code

**Comment by mshima** [comment]
I think we should generate Types from command instead of reworking types see https://github.com/jhipster/generator-jhipster/pull/27133.

**Comment by Tcharl** [comment]
I got it already: command act as the single source of truth.
Still, at on point in time during the generation execution lifecycle:

1. Commands will be [merged with defaults when no answer](https://github.com/jhipster/generator-jhipster/pull/27133/files#diff-e5784763ddfff3ac6ecc0d138a92136255f3de1cd5f9bd856a00b6c3e04fde65): here, the objective of this PR in the long term will be to get rid of the 'jhipsterconfigWithDefault' in favor of default values introduced in the respective `<generator>/command.ts`.
2. Commands results will be injected in an object representing all the answers of a user for generator prompts: [configure part of the command](https://github.com/jhipster/generator-jhipster/blob/09db3d8991b0c77d157a1558a646968546ed76eb/generators/spring-boot/command.ts#L160): the goal of this PR is to materialize the resulting typescript object (let's call it `GeneratorXConfig`) and all its attributes (an object per generator), instead of using `jhipsterConfig` which is the aggregate of the GeneratorConfigs of all the generators and which should only exist temporary.
4. These objects will then be enriched with derived properties `EnrichedGeneratorXConfig`.
3. These EnrichedGeneratorConfigs are then classified: platforms ones, application ones and entities-related: each generator Config will be classified in one of these three category (PlatformConfig, ApplicationConfig, EntityConfig). -> I'm not sure this step should happen, but I guess that sometimes you'll need an aggregate to produce a full template, your opinion?
5. Finally, these EnrichedConfigs will be used in preprocess, process and postprocess: here, the developer will definitely be happy to get autocomplete and know at which phase he can have access to which attribute of this config in his template, which is why I started the PR.

I know it's far from being finished yet and that it is quite an adventure to get rid of the current pains related:
1. Knowing configuration attributes available at each lifecycle phase
2. Modular generator architecture not centralized on 'big aggregate' objects which would help blueprints
3. JDL vs Cli gaps and synchronization.

But we have to start from something and this is a small step to start from

File: generators/spring-cloud/generators/gateway/jdl/jdl-routes-option.spec.ts
Comment (mshima) [inline generators/spring-cloud/generators/gateway/jdl/jdl-routes-option.spec.ts:1]: Unit test needs a minimum reproduction, if we provide a complete sample it may break with unrelated changes.

File: jdl/converters/json-to-jdl-converter.spec.ts
Comment (mshima) [inline jdl/converters/json-to-jdl-converter.spec.ts:1]: Unit test needs a minimum reproduction, if we provide a complete sample it may break with unrelated changes.

File: generators/base-application/types.d.ts
Comment (mshima) [inline generators/base-application/types.d.ts:1]: These properties are available in app/command.
We should extract from there.

File: generators/client/types.d.ts [hunk 1]
```diff
@@ -10,12 +10,17 @@ type ClientFrameworkType = ['no', 'angular', 'react', 'vue', 'svelte'];
 
 type ClientFrameworkApplication = OptionWithDerivedProperties<'clientFramework', ClientFrameworkType>;
 
+export type PersistedClientApplication = {
```
Comment (mshima) [inline generators/client/types.d.ts:13]: Persisted types are the ExportStoragePropertiesFromCommand types.

File: generators/client/types.d.ts [hunk 2]
```diff
@@ -10,12 +10,17 @@ type ClientFrameworkType = ['no', 'angular', 'react', 'vue', 'svelte'];
...
+  withAdminUi: boolean;
```
Comment (mshima) [inline generators/client/types.d.ts:14]: It’s already available in the bellow type.

File: jdl/converters/types.d.ts
Comment (mshima) [inline jdl/converters/types.d.ts:1]: Default types should be dropped from converters and replaced with generics.
We should not try to export or integrate them.

File: generators/base-core/generator.ts [hunk 3]
```diff
@@ -68,6 +68,7 @@ import { getGradleLibsVersionsProperties } from '../gradle/support/dependabot-gr
 import { dockerPlaceholderGenerator } from '../docker/utils.js';
 import { getConfigWithDefaults } from '../../jdl/index.js';
 import { extractArgumentsFromConfigs } from '../base/internal/command.js';
+import { JSONGeneratorJhipsterContent } from '../../jdl/converters/types.js';
```
Comment (mshima) [inline generators/base-core/generator.ts:71]: We should not mix jdl with generator types.
To add a type to jhipsterConfig a new type should be created join storage types from every generator.

**Comment by Tcharl** [comment]
Gonna retry with a smaller scope

File: generators/base-application/types.d.ts
Comment (Tcharl) [reply generators/base-application/types.d.ts:1]: Still, at one point it has to be materialized to an object
